File: create_project/create_project.py
#!.env/bin/python3

import os
import sqlite3
import subprocess
import sys

from .helper import Helper


class CreateProject:

    # Path to the database
    conf_dir = os.environ["HOME"] + "/.config/create_project/"
    template_path = conf_dir + "templates/"
    script_path = conf_dir + "scripts/"
    db = conf_dir + "db/cp.sqlite3"

    def __init__(self, lang, project_name):
        if not type(lang) is str:
            raise TypeError("Language name must be a string.")
        if not type(project_name) is str:
            raise TypeError("Project name must be a string.")
        self.lang = lang
        self.project_name = project_name
        self.project_id = None
        self.cwd = os.getcwd()
        self.conn = None
        self.cur = None

    def run(self):
        # Connect to database
        with sqlite3.connect(CreateProject.db) as self.conn:
            if not self.conn:
                err_msg = Helper.format_err_msg("Could not connect to database.")
                print(err_msg)
                return 2

            # Create cursor
            self.cur = self.conn.cursor()

            if not self.cur:
                return 3

            # Check if provided language is supported
            if not self._is_extension_supported():
                return 4

            # Create the project folder
            if not self._create_project_folder():
                return 5

            # Change directory into new project folder
            os.chdir(self.project_name)

            # Create language specific sub folders
            if not self._create_sub_folders():
                return 6

            # Create language specific files
            if not self._create_files():
                return 7

            # Copy template files
            if not self._copy_templates():
                return 8

            # Run custom scripts
            if not self._run_scripts():
                return 9

            # Cd back to old cwd
            os.chdir(self.cwd)
        return 0

    def _is_extension_supported(self):
        """ Check if the provided file extension is supported. """

        try:
            self.cur.execute(
                """
                SELECT
                    project_id,
                    extension
                FROM
                    file_extension
                WHERE
                    extension=?
                """,
                (self.lang,),
            )

            lang = self.cur.fetchone()

        except sqlite3.Error as err:
            print(err)
            return False

        if lang:
            self.project_id = lang[0]
            return True

        # Language not supported
        extensions = self.cur.execute("SELECT extension FROM file_extension").fetchall()

        err_msg = Helper.format_err_msg(
            "You have to specify a supported file extension.",
            ("Currently supported languages: " + str(extensions)),
        )

        print(err_msg)
        return False

    def _create_project_folder(self):
        """ Create the main project directory. """

        print("> Creating project folder...")
        try:
            subprocess.run(["mkdir", self.project_name], timeout=10.0, check=True)
        except (subprocess.CalledProcessError, TimeoutError) as err:
            print(err)
            return False
        return True

    def _create_sub_folders(self):
        """ Create sub folders depending on the specified language. """

        print("> Creating subfolders...")

        try:
            self.cur.execute(
                """
                SELECT
                    target_path
                FROM
                    project_folder
                WHERE
                    (project_id=? OR project_id IS NULL)
                    AND template_name IS NULL
                """,
                (self.project_id,),
            )

            subfolders = self.cur.fetchall()

            for subfolder in subfolders:
                subprocess.run(["mkdir", "-p", subfolder[0]], timeout=10.0, check=True)

        except (sqlite3.Error, subprocess.CalledProcessError, TimeoutError) as err:
            print(err)
            return False
        return True

    def _create_files(self):
        """ Create language specific files. """

        print("> Creating files...")

        try:
            self.cur.execute(
                """
                SELECT
                    target_path
                FROM
                    project_file
                WHERE
                    (project_id=? OR project_id IS NULL)
                    AND template_name IS NULL
                """,
                (self.project_id,),
            )

            files = self.cur.fetchall()

            for file in files:
                subprocess.run(["touch", file[0]], timeout=10.0, check=True)

        except (sqlite3.Error, subprocess.CalledProcessError, TimeoutError) as err:
            print(err)
            return False
        return True

    def _copy_templates(self):
        """ Create language specific files. """

        print("> Copying templates...")

        try:
            self.cur.execute(
                """
                SELECT
                    target_path, template_name
                FROM
                    project_file
                WHERE
                    (project_id=? OR project_id IS NULL)
                    AND template_name IS NOT NULL
                """,
                (self.project_id,),
            )

            template_files = self.cur.fetchall()

            self.cur.execute(
                """
                SELECT
                    target_path, template_name
                FROM
                    project_folder
                WHERE
                    (project_id=? OR project_id IS NULL)
                    AND template_name IS NOT NULL
                """,
                (self.project_id,),
            )

            template_folders = self.cur.fetchall()

            for file in template_files:
                template = CreateProject.template_path + file[1]
                target = file[0]
                subprocess.run(["cp", template, target], timeout=30.0, check=True)

            for folder in template_folders:
                template = CreateProject.template_path + folder[1]
                target = folder[0]
                subprocess.run(["cp", "-r", template, target], timeout=30.0, check=True)

        except (sqlite3.Error, subprocess.CalledProcessError, TimeoutError) as err:
            print(err)
            return False
        return True

    def _run_scripts(self):
        """ Run custom scripts. """

        try:
            self.cur.execute(
                """
                SELECT
                    script_name,
                    run_as_sudo
                FROM
                    project_script
                WHERE
                    project_id is NULL
                    OR project_id=?
                ORDER BY project_id DESC
                """,
                (self.project_id,),
            )

            scripts = self.cur.fetchall()

            for script in scripts:
                script_path = self.script_path + script[0]
                run_as_sudo = bool(script[1])

                if run_as_sudo:
                    subprocess.run(["sudo", script_path], timeout=60.0, check=True)
                else:
                    subprocess.run([script_path], timeout=60.0, check=True)

        except (sqlite3.Error, subprocess.CalledProcessError, TimeoutError) as err:
            print(err)
            return False
        return True

File: internal/app/createproject/createproject.go
package createproject

import (
	"database/sql"
	"fmt"
	"os"
	"os/exec"

	// Import sqlite3 driver
	_ "github.com/mattn/go-sqlite3"

	"github.com/otiai10/copy"
)

// Setup contains necessary informations for the creation of a project.
// Owd is the Origin Working Directory.
type Setup struct {
	Owd          string
	ConfigDir    string
	DatabaseName string
	Extension    string
	templatesDir string
	scriptsDir   string
	dbDir        string
	db           *sql.DB
	projectID    string
}

// Run starts the creation of a project
func (setup *Setup) Run() error {
	// Set dirs
	setup.dbDir = setup.ConfigDir + "db/"
	setup.templatesDir = setup.ConfigDir + "templates/"
	setup.scriptsDir = setup.ConfigDir + "scripts/"

	// Connect to database
	db, err := sql.Open("sqlite3", setup.dbDir+setup.DatabaseName)
	if err != nil {
		return err
	}
	setup.db = db

	// Check if extension is supported
	err = setup.isExtensionSupported()
	if err != nil {
		return err
	}
	return nil
}

// Stop cleanly stops the running Setup instance.
// Currently it's only closing its open database connection.
func (setup *Setup) Stop() {
	// Close database connection
	if setup.db != nil {
		setup.db.Close()
	}
}

// isExtensionSupported checks if the given extension is found in the database.
// Returns nil if found, returns error if not found
func (setup *Setup) isExtensionSupported() error {
	stmt, err := setup.db.Prepare("SELECT project_id FROM file_extension WHERE extension = ?")
	if err != nil {
		return err
	}
	defer stmt.Close()
	var id string
	err = stmt.QueryRow(setup.Extension).Scan(&id)
	if err != nil {
		return err
	}
	setup.projectID = id
	return nil
}

// Project struct represents a project that will be build.
// Containing information about project name and extension.
// The setup struct includes information about config paths and a open database connection.
type Project struct {
	id   string
	Name string
	Data *Setup
}

// Create starts the creation of a project.
// Returns an error on failure. Returns nil on success.
func (project *Project) Create() error {
	// Create the project folder
	fmt.Println("> Creating project folder...")
	err := project.createProjectFolder()
	if err != nil {
		return err
	}

	// Chdir into the new project folder and defer chdir back to old cwd
	err = os.Chdir(project.Name)
	if err != nil {
		return err
	}
	defer os.Chdir(project.Data.Owd)

	// Create subfolders
	fmt.Println("> Creating subfolders...")
	err = project.createSubFolders()
	if err != nil {
		return err
	}

	// Create files
	fmt.Println("> Creating files...")
	err = project.createFiles()
	if err != nil {
		return err
	}

	// Copy templates
	fmt.Println("> Copying templates...")
	err = project.copyTemplates()
	if err != nil {
		return err
	}

	// Run scripts
	fmt.Println("> Running scripts...")
	err = project.runScripts()
	if err != nil {
		return err
	}
	return nil
}

// createProjectFolder tries to create the main project folder.
// Returns an error on failure.
func (project *Project) createProjectFolder() error {
	err := os.Mkdir(project.Name, os.ModePerm)
	if err != nil {
		return err
	}
	return nil
}

// createSubFolders queries all subfolder from the database related to the projectId.
// Tries to create all of the subfolders in the projectfolder.
// Returns error on failure. Returns nil on success.
func (project *Project) createSubFolders() error {
	// Query subfolders
	stmt, err := project.Data.db.Prepare("SELECT target_path FROM project_folder WHERE (project_id = ? OR project_id IS NULL) AND template_name IS NULL")
	if err != nil {
		return err
	}
	defer stmt.Close()

	subFolders, err := stmt.Query(project.Data.projectID)
	if err != nil {
		return err
	}
	defer subFolders.Close()

	// Create subfolders
	for subFolders.Next() {
		var subFolder string
		err = subFolders.Scan(&subFolder)
		if err != nil {
			return err
		}
		err = os.MkdirAll(subFolder, os.ModePerm)
		if err != nil {
			return err
		}
	}
	return nil
}

// createFiles queries all files from the database related to the projectId.
// Tries to create all of the files in the projectfolder.
// Returns error on failure. Returns nil on success.
func (project *Project) createFiles() error {
	// Query files
	stmt, err := project.Data.db.Prepare("SELECT target_path FROM project_file WHERE (project_id = ? OR project_id IS NULL) AND template_name IS NULL")
	if err != nil {
		return err
	}
	defer stmt.Close()

	files, err := stmt.Query(project.Data.projectID)
	if err != nil {
		return err
	}
	defer files.Close()

	// Create files
	for files.Next() {
		var file string
		err = files.Scan(&file)
		if err != nil {
			return err
		}
		f, err := os.OpenFile(file, os.O_RDONLY|os.O_CREATE, os.ModePerm)
		if err != nil {
			return err
		}
		f.Close()
	}
	return nil
}

// copyTemplates queries all templates from the database related to the projectId.
// Tries to copy all of the templates into the projectfolder.
// Returns error on failure. Returns nil on success.
func (project *Project) copyTemplates() error {
	// Query template folders
	stmt, err := project.Data.db.Prepare(
		"SELECT target_path, template_name FROM project_folder WHERE (project_id = ? OR project_id IS NULL) AND template_name IS NOT NULL")
	if err != nil {
		return err
	}
	defer stmt.Close()

	folders, err := stmt.Query(project.Data.projectID)
	if err != nil {
		return err
	}
	defer folders.Close()

	// Copy template files
	for folders.Next() {
		var target, src string
		err = folders.Scan(&target, &src)
		if err != nil {
			return err
		}
		src = project.Data.templatesDir + src
		err := copy.Copy(src, target)
		if err != nil {
			return err
		}
	}

	// Query template files
	stmt, err = project.Data.db.Prepare(
		"SELECT target_path, template_name FROM project_file WHERE (project_id = ? OR project_id IS NULL) AND template_name IS NOT NULL")
	if err != nil {
		return err
	}

	files, err := stmt.Query(project.Data.projectID)
	if err != nil {
		return err
	}
	defer files.Close()

	// Copy template files
	for files.Next() {
		var target, src string
		err = files.Scan(&target, &src)
		if err != nil {
			return err
		}
		src = project.Data.templatesDir + src
		err := copy.Copy(src, target)
		if err != nil {
			return err
		}
	}

	return nil
}

// runScripts queries all scripts from the database related to the projectId.
// Tries to execute all scripts.
// Returns error on failure. Returns nil on success.
func (project *Project) runScripts() error {
	// Query scripts
	stmt, err := project.Data.db.Prepare("SELECT script_name, run_as_sudo FROM project_script WHERE project_id is NULL OR project_id = ? ORDER BY project_id DESC")
	if err != nil {
		return err
	}
	defer stmt.Close()

	scripts, err := stmt.Query(project.Data.projectID)
	if err != nil {
		return err
	}
	defer scripts.Close()

	// Create scripts
	for scripts.Next() {
		var script string
		var runAsSudo int
		err = scripts.Scan(&script, &runAsSudo)
		if err != nil {
			return err
		}
		script = project.Data.scriptsDir + script
		err = exec.Command(script).Run()
		if err != nil {
			return err
		}
	}
	return nil
}
